src/Database/Driver.php

```diff
@@ -463,44 +464,45 @@ public function quote(string $value, int $type = PDO::PARAM_STR): string
     }
 
     /**
-     * Returns a closure that will be used to transform a passed Query object.
-     * This function, in turn, will return an instance of a Query object that has been
-     * transformed to accommodate any specificities of the SQL dialect in use.
+     * Transform the query to accommodate any specificities of the SQL dialect in use.
+     *
+     * It will also quote the identifiers if auto quoting is enabled.
      *
-     * @param string $type the type of query to be transformed
-     * (select, insert, update, delete)
-     * @return \Closure
+     * @param \Cake\Database\Query $query Query to transform.
+     * @return \Cake\Database\Query
      */
-    public function queryTranslator(string $type): Closure
+    protected function transformQuery(Query $query): Query
     {
-        return function ($query) use ($type) {
-            if ($this->isAutoQuotingEnabled()) {
-                $query = (new IdentifierQuoter($this))->quote($query);
-            }
+        $query = match (true) {
+            $query instanceof SelectQuery => $this->_selectQueryTranslator($query),
+            $query instanceof InsertQuery => $this->_insertQueryTranslator($query),
+            $query instanceof UpdateQuery => $this->_updateQueryTranslator($query),
+            $query instanceof DeleteQuery => $this->_deleteQueryTranslator($query),
+            default =>
+                throw new InvalidArgumentException(sprintf(
+                    'Instance of SelectQuery, UpdateQuery, InsertQuery, DeleteQuery expected. Found `%s` instead.',
```

Comment:
Can't this be typed now?

Reply:
The is required because the method argument `$query` is typed using the parent class `Query`.

Reply:
Yea, I was wondering if we should type the argument to SelectQuery, etc types.

Reply:
Not really. That would require changing the type from `Query` to union of the specific query classes throughout the package, otherwise static analyzers will complain about passing a parent type where a specifx type is required.